src/coreclr/nativeaot/Runtime/amd64/GcProbe.S

```diff
@@ -109,88 +91,43 @@
         pop rdx
         pop rax
 
-        //
         // Fix the stack by pushing the original return address
-        //
         mov         rcx, [r11 + OFFSETOF__Thread__m_pvHijackedReturnAddress]
         push        rcx
 
+        // Fetch the return address flags
         mov         rcx, [r11 + OFFSETOF__Thread__m_uHijackedReturnValueFlags]
 
-        ClearHijackState
-.endm
-
-//////////////////////////////////////////////////////////////////////////////////////////////////////////////
-//
-// RhpWaitForGCNoAbort -- rare path for WaitForGCCompletion
-//
-//
-// INPUT: RDI: transition frame
-//
-// TRASHES: RCX, RDI, R8, R9, R10, R11
-//
-//////////////////////////////////////////////////////////////////////////////////////////////////////////////
-NESTED_ENTRY RhpWaitForGCNoAbort, _TEXT, NoHandler
-        END_PROLOGUE
-
-        mov         rdx, [rdi + OFFSETOF__PInvokeTransitionFrame__m_pThread]
-
-        test        dword ptr [rdx + OFFSETOF__Thread__m_ThreadStateFlags], TSF_DoNotTriggerGc
-        jnz         Done
-
-        // passing transition frame pointer in rdi
-        call        C_FUNC(RhpWaitForGC2)
-
-Done:
-        ret
-
-NESTED_END RhpWaitForGCNoAbort, _TEXT
-
-//
-// Set the Thread state and wait for a GC to complete.
-//
-// Register state on entry:
-//  RBX: thread pointer
-//
-// Register state on exit:
-//  RBX: thread pointer
-//  All other registers trashed
-//
-
-.macro WaitForGCCompletion
-        test        dword ptr [rbx + OFFSETOF__Thread__m_ThreadStateFlags], TSF_SuppressGcStress + TSF_DoNotTriggerGc
-        jnz         LOCAL_LABEL(NoWait)
-
-        mov         rdi, [rbx + OFFSETOF__Thread__m_pDeferredTransitionFrame]
-        call        C_FUNC(RhpWaitForGCNoAbort)
-LOCAL_LABEL(NoWait):
-
+        // Clear hijack state
+        xor         r9, r9
+        mov         [r11 + OFFSETOF__Thread__m_ppvHijackedReturnAddressLocation], r9
+        mov         [r11 + OFFSETOF__Thread__m_pvHijackedReturnAddress], r9
+        mov         [r11 + OFFSETOF__Thread__m_uHijackedReturnValueFlags], r9
 .endm
 
-//
-//
 //
 // GC Probe Hijack target
 //
-//
-
 NESTED_ENTRY RhpGcProbeHijack, _TEXT, NoHandler
         END_PROLOGUE
         FixupHijackedCallstack
+
+        test        dword ptr [C_VAR(RhpTrapThreads)], TrapThreadsFlags_TrapThreads
+        jnz         LOCAL_LABEL(DoRhpGcProbe)
+        ret
+
+LOCAL_LABEL(DoRhpGcProbe):
         or          ecx, DEFAULT_FRAME_SAVE_FLAGS  + PTFF_SAVE_RAX + PTFF_SAVE_RDX
         jmp         C_FUNC(RhpGcProbe)
 NESTED_END RhpGcProbeHijack, _TEXT
 
 NESTED_ENTRY RhpGcProbe, _TEXT, NoHandler
```

Comment:
`RhpGcProbe` does not match what the funciton does now that `RhpTrapThreads` check was removed from it. Should it be called `RhpWaitForGC` instead?

Reply:
RhpGcProbe was not the best name. 
The actual purpose of the function is to set up the stackwalk-enabling frame around the `RhpWaitForGC2`, which does the actual waiting. It is hard to capture that in couple words.

`RhpWaitForGC` seems a bit better. The function does typically result in waiting for GC.